File: 28. Reorder list.py
class Node:
    def __init__(self,data): self.data=data; self.next=None

def reorder(head):
    if not head: return
    vals=[]; cur=head
    while cur:
        vals.append(cur)
        cur=cur.next
    i=0; j=len(vals)-1
    while i<j:
        vals[i].next=vals[j]; i+=1
        if i==j: break
        vals[j].next=vals[i]; j-=1
    vals[i].next=None

def print_list(h):
    cur=h
    while cur:
        print(cur.data, end=' '); cur=cur.next
    print()

if __name__=='__main__':
    head=Node(1); head.next=Node(2); head.next.next=Node(3); head.next.next.next=Node(4); head.next.next.next.next=Node(5)
    reorder(head)
    print_list(head)

File: 28. Reorder list.cpp
#include <iostream>
#include <vector>
using namespace std; struct Node{int data; Node* next; Node(int d):data(d),next(nullptr){}
};
void reorder(Node* head){ if(!head) return; vector<Node*> v; Node* cur=head; while(cur){ v.push_back(cur); cur=cur->next; } int i=0,j=v.size()-1; while(i<j){ v[i]->next=v[j]; i++; if(i==j) break; v[j]->next=v[i]; j--; } v[i]->next=nullptr; }
void print(Node* h){ while(h){ cout<<h->data<<" "; h=h->next; } cout<<endl; }
int main(){ Node* head=new Node(1); head->next=new Node(2); head->next->next=new Node(3); head->next->next->next=new Node(4); head->next->next->next->next=new Node(5); reorder(head); print(head); return 0; }
